Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -746,6 +783,35 @@ public class ProgramBuilder {
                     add(instr)
                 }
             }
+
+            // When we encounter a block begin:
+            // 1. We ensure that the context being opened removes atleast one required context
+            // 2. The required context is not empty
+            // 3. The required context is not equal to self.context (we don't want to capture blocks don't open atleast one requried context)
+            if instr.isBlockBegin {
+                var requiredContext = requiredContextStack.popLast()!
+                if requiredContext.subtracting(instr.op.contextOpened) != requiredContext && requiredContext != .empty && requiredContext != self.context {
+                    requiredContextStack.append(requiredContext)
+                    add(instr)
+                    requiredContext = requiredContextStack.popLast()!
+                } 
+                requiredContext = requiredContext.subtracting(instr.op.contextOpened)
+
+                // We must have atleast one context on the stack
+                if requiredContextStack.count < 1 {
```

Comment:
I have a testcase for this scenario: `testBeginForSplicing`.
So when we capture the `ForLoop` block and as a result `Await`, as we walk back up the instruction stack we encounter `BeginAsyncFunction` which opens a context that we need for `Await` so we capture the block begin and end, without the block contents.

Reply:
Hmm yeah, I think the example needs to be a bit more complicated than what I wrote above. Does this work:
```
v1 <- LoadInt 42
v2 <- BeginFunctionDefinition
  BeginForLoop
    v3 <- LoadArguments
    Reassign v1, v3
  EndForLoop
EndFunctionDefinition
CallFunction v2, v1
```
Then put the builder into `.conservative` mode and start the splicing at `CallFunction`. Then it'll push a new context onto the `requiredContextStack` when finding the `EndFunctionDefinition` and again for `EndForLoop`, then decide it needs to capture the `Reassign`, then take the `LoadArguments` as well, at which point the top of the context stack is `[.function]`. When it then finds the `BeginForLoop`, from what I can tell, the current code would just drop the top of the stack and not take the `BeginForLoop` (which is correct), but then it'll also not take the `BeginFunctionDefinition` later on, which is incorrect. So instead of dropping the context, I think it needs to merge it with the one below.

Reply:
Ah no, the above example doesn't work, my bad, this is what I meant:
```
v1 <- LoadInt 42
BeginFunctionDefinition
  BeginForLoop
    v3 <- LoadArguments
    Reassign v1, v3
  EndForLoop
EndFunctionDefinition
CreateArray [v1]
```
Then start at `CreateArray`

Reply:
Good catch! this does lead to an invalid program. I'll debug this further and push a fix.

Reply:
Cool, glad it worked :D Yeah I think the fix is to basically merge the top two states when you reach an `BeginX` instead of just removing the top entry. Here, merging means subtracting the states opened by the `BeginX` if it is kept, then combining (unioning) the result with the other state. Unless there is only one entry in the stack, in which case you'd just keep the top state after the subtraction, like you currently do.

Reply:
This is what currently happens when we encounter `BeginFor`:
stack: [[.script], [.empty], [.script, .function]]
we pop the stack to retrieve (`requiredContext`): [.script, .function]
context opened by BeginFor: [.scrtipt, .loop]
subtracting `contextOpened` from `requiredContext` we get: [.function]
and since the result is not equal to `requiredContext` we end up capturing the `BeginFor` instruction. So now I think the result of the subtraction should be a subset of `contextOpened` in order to keep the instruction.

Reply:
This should now be resolved with the latest commit